lib/pbench/server/indexing_tarballs.py

```diff
@@ -436,9 +439,56 @@ def sighup_handler(*args):
                                         if tb_res.success
                                         else States.QUARANTINED
                                     )
+
+                                    # In case this was a re-index, remove the
+                                    # REINDEX tag.
                                     Metadata.remove(dataset, Metadata.REINDEX)
+
+                                    # Because we're on the `finally` path, we
+                                    # can get here without a PbenchTarBall
+                                    # object, so don't try to write an index
+                                    # map if there is none.
+                                    if ptb:
+                                        # A pbench-index --tool-data follows a
+                                        # pbench-index and generates only the
+                                        # tool-specific documents: we want to
+                                        # merge that with the existing document
+                                        # map. On the other hand, a re-index
+                                        # should replace the entire index. We
+                                        # accomplish this by overwriting each
+                                        # duplicate index key separately.
+                                        try:
+                                            meta = Metadata.get(
+                                                dataset, Metadata.INDEX_MAP
+                                            )
+                                            map = json.loads(meta.value)
+                                            map.update(ptb.index_map)
+                                            meta.value = json.dumps(map)
+                                            meta.update()
+                                        except MetadataNotFound:
+                                            Metadata.create(
+                                                dataset=dataset,
+                                                key=Metadata.INDEX_MAP,
+                                                value=json.dumps(ptb.index_map),
+                                            )
```

Comment:
It's critical to publish, because I need the document map.

And, yeah, the whole module should be refactored; I'm not going to deal with that for this PR, but maybe I can at least avoid adding to it by pulling this code out into a separate method. I probably should have done that when it got above a handful of lines; but it didn't happen all at once.